File: shared/src/commonMain/kotlin/com/appautopsy/analysis/core/SignerHistory.kt
package com.appautopsy.analysis.core

/**
 * Shared cert-change detection (F12 signal 3 + F13).
 *
 * "Same package name, different signer than last time" is a textbook
 * repackaging indicator, and it is the one signal that needs no brand data at
 * all. It lives here so repackaging detection and the version-diff tracker
 * consume one implementation rather than two that can diverge.
 */
object SignerHistory {

    /**
     * True when a prior cert exists and none of the current certs match it.
     *
     * An empty side means "no comparison possible", which answers false:
     * absence of data is never treated as evidence of an attack.
     */
    fun hasSignerChanged(
        priorCertSha256: String?,
        currentCertsSha256: List<String>,
    ): Boolean {
        val prior = priorCertSha256?.trim()?.lowercase()
        if (prior.isNullOrEmpty() || currentCertsSha256.isEmpty()) return false

        val current = currentCertsSha256
            .filter { it.isNotBlank() }
            .map { it.trim().lowercase() }
            .toSet()

        return prior !in current
    }
}

File: backend/app/core/signer_history.py
"""Shared cert-change detection helper (F12 signal 3 + F13).

Compares current signing certificate fingerprints against historical fingerprints
for the same package.
"""

from __future__ import annotations


def has_signer_changed(
    prior_cert_sha256: str | None,
    current_certs_sha256: list[str],
) -> bool:
    """Returns True if a prior cert exists and none of the current certs match it."""
    if not prior_cert_sha256 or not current_certs_sha256:
        return False

    prior = prior_cert_sha256.strip().lower()
    current = {c.strip().lower() for c in current_certs_sha256 if c}

    return prior not in current
